File: ios/Algo/Features/Chat/ViewModels/ChatViewModel.swift
// ChatViewModel.swift
// Algo
//
// Main ViewModel for the chat interface. Manages message history, sending/receiving,
// WebSocket streaming, action confirmation, and suggested actions.

import Foundation
import Combine

// MARK: - ChatViewModel

@MainActor
final class ChatViewModel: ObservableObject {

    // MARK: - Published Properties

    /// The list of messages in the current conversation, ordered chronologically.
    @Published private(set) var messages: [Message] = []

    /// Whether a message send/receive operation is in progress.
    @Published private(set) var isLoading = false

    /// Whether the assistant is actively streaming a response via WebSocket.
    @Published private(set) var isStreaming = false

    /// The current error message to display, if any.
    @Published var error: String?

    /// Suggested follow-up actions returned by the last assistant message.
    @Published private(set) var suggestedActions: [SuggestedAction]?

    /// The current conversation ID. Nil for a new conversation.
    @Published private(set) var conversationId: String?

    // MARK: - Dependencies

    private let chatService: ChatServiceProtocol
    private let conversationRepository: ConversationRepositoryProtocol
    private let webSocketClient: WebSocketClient

    // MARK: - Internal State

    /// Accumulates streaming text chunks for the in-progress assistant message.
    private var streamingContent = ""

    /// The ID of the message currently being streamed.
    private var streamingMessageId: String?

    /// Combine cancellables for WebSocket event subscriptions.
    private var cancellables = Set<AnyCancellable>()

    /// Pagination state for loading older messages.
    private var currentPage = 1
    private var hasMoreMessages = false
    private var isLoadingMore = false

    // MARK: - Initialization

    /// Creates a new ChatViewModel.
    ///
    /// - Parameters:
    ///   - chatService: The service for sending messages and confirming actions.
    ///   - conversationRepository: The repository for loading message history.
    ///   - webSocketClient: The WebSocket client for real-time streaming.
    ///   - conversationId: An existing conversation ID to resume, or nil for new.
    init(
        chatService: ChatServiceProtocol,
        conversationRepository: ConversationRepositoryProtocol,
        webSocketClient: WebSocketClient,
        conversationId: String? = nil
    ) {
        self.chatService = chatService
        self.conversationRepository = conversationRepository
        self.webSocketClient = webSocketClient
        self.conversationId = conversationId
    }

    // MARK: - Public Methods

    /// Sends a text message to the assistant.
    ///
    /// Adds the user message to the local list immediately for responsiveness,
    /// then calls the chat API. On success, the assistant's response is appended.
    ///
    /// - Parameter text: The user's message text.
    func sendMessage(_ text: String) {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }

        // Add user message locally for instant feedback
        let userMessage = Message(
            role: .user,
            content: trimmed,
            inputType: "text"
        )
        messages.append(userMessage)
        suggestedActions = nil
        error = nil

        isLoading = true

        Task {
            do {
                let response = try await chatService.sendMessage(
                    conversationId: conversationId,
                    message: trimmed,
                    inputType: "text"
                )

                handleChatResponse(response)
            } catch {
                handleError(error)
            }

            isLoading = false
        }
    }

    /// Executes a suggested action by mapping it to a message send.
    ///
    /// Special actions like CONFIRM_ACTION and CANCEL_ACTION are routed to
    /// the confirmation flow. All others are sent as a regular user message
    /// using the action's label text.
    ///
    /// - Parameter action: The suggested action to execute.
    func executeSuggestedAction(_ action: SuggestedAction) {
        switch action.action {
        case "CONFIRM_ACTION":
            if let actionId = action.params?["action_id"]?.stringValue {
                confirmAction(confirmationId: actionId, confirmed: true)
            }

        case "CANCEL_ACTION":
            if let actionId = action.params?["action_id"]?.stringValue {
                confirmAction(confirmationId: actionId, confirmed: false)
            }

        default:
            sendMessage(action.messageText)
        }
    }

    /// Confirms or cancels a pending action.
    ///
    /// - Parameters:
    ///   - confirmationId: The ID of the confirmation to respond to.
    ///   - confirmed: Whether the action is confirmed (true) or cancelled (false).
    func confirmAction(confirmationId: String, confirmed: Bool) {
        guard let conversationId else {
            error = "No active conversation for confirmation."
            return
        }

        isLoading = true
        error = nil

        Task {
            do {
                let response = try await chatService.confirmAction(
                    conversationId: conversationId,
                    confirmationId: confirmationId,
                    confirmed: confirmed
                )

                handleChatResponse(response)
            } catch {
                handleError(error)
            }

            isLoading = false
        }
    }

    /// Connects the WebSocket and begins listening for streaming events.
    ///
    /// Call this when the chat view appears to enable real-time streaming
    /// responses, typing indicators, and alerts.
    func connectWebSocket() {
        webSocketClient.connect()
        subscribeToWebSocketEvents()
    }

    /// Disconnects the WebSocket connection.
    ///
    /// Call this when the chat view disappears to conserve resources.
    func disconnectWebSocket() {
        webSocketClient.disconnect()
        cancellables.removeAll()
    }

    /// Loads older messages for the current conversation (pagination).
    func loadMoreMessages() async {
        guard let conversationId,
              hasMoreMessages,
              !isLoadingMore else { return }

        isLoadingMore = true

        do {
            let nextPage = currentPage + 1
            let olderMessages = try await conversationRepository.getMessages(
                conversationId: conversationId,
                page: nextPage,
                limit: 50
            )

            // Prepend older messages to the beginning
            messages.insert(contentsOf: olderMessages, at: 0)
            currentPage = nextPage
            // If fewer messages returned than requested, no more pages
            hasMoreMessages = olderMessages.count >= 50
        } catch {
            self.error = userFriendlyMessage(for: error)
        }

        isLoadingMore = false
    }

    /// Loads the initial message history for an existing conversation.
    func loadInitialMessages() async {
        guard let conversationId else { return }

        isLoading = true

        do {
            let loadedMessages = try await conversationRepository.getMessages(
                conversationId: conversationId,
                page: 1,
                limit: 50
            )

            messages = loadedMessages
            currentPage = 1
            hasMoreMessages = loadedMessages.count >= 50

            // Set suggested actions from the last assistant message
            if let lastAssistant = messages.last(where: { $0.isAssistant }) {
                suggestedActions = lastAssistant.suggestedActions
            }
        } catch {
            self.error = userFriendlyMessage(for: error)
        }

        isLoading = false
    }

    /// Clears the conversation state for starting a new chat.
    func startNewConversation() {
        messages.removeAll()
        conversationId = nil
        suggestedActions = nil
        error = nil
        currentPage = 1
        hasMoreMessages = false
        streamingContent = ""
        streamingMessageId = nil
        isStreaming = false
    }

    // MARK: - Private Methods

    /// Processes a successful chat API response.
    private func handleChatResponse(_ response: ChatAPIResponse) {
        // Update conversation ID (may be assigned by server for new conversations)
        conversationId = response.conversationId

        // Convert the API response to a Message and append it
        let assistantMessage = response.toMessage()
        messages.append(assistantMessage)

        // Update suggested actions
        suggestedActions = response.response.suggestedActions
    }

    /// Subscribes to WebSocket events for real-time streaming.
    private func subscribeToWebSocketEvents() {
        webSocketClient.events
            .receive(on: DispatchQueue.main)
            .sink { [weak self] event in
                self?.handleWebSocketEvent(event)
            }
            .store(in: &cancellables)
    }

    /// Routes individual WebSocket events to the appropriate handler.
    private func handleWebSocketEvent(_ event: WebSocketEvent) {
        guard let conversationId else { return }

        switch event {
        case .typing(let eventConversationId):
            guard eventConversationId == conversationId else { return }
            if !isStreaming {
                isStreaming = true
                streamingContent = ""
            }

        case .streamChunk(let eventConversationId, let chunk):
            guard eventConversationId == conversationId else { return }
            handleStreamChunk(chunk)

        case .streamEnd(let eventConversationId, let messageId):
            guard eventConversationId == conversationId else { return }
            handleStreamEnd(messageId: messageId)

        case .connected:
            break

        case .disconnected:
            if isStreaming {
                finalizeStreamingMessage()
            }

        case .alert, .metricUpdate:
            break
        }
    }

    /// Accumulates a streaming text chunk into the in-progress assistant message.
    private func handleStreamChunk(_ chunk: String) {
        isStreaming = true
        streamingContent += chunk

        if streamingMessageId == nil {
            let placeholderId = "streaming_\(UUID().uuidString.prefix(8))"
            streamingMessageId = placeholderId

            let streamingMessage = Message(
                id: placeholderId,
                role: .assistant,
                content: streamingContent
            )
            messages.append(streamingMessage)
        } else if let index = messages.lastIndex(where: { $0.id == streamingMessageId }) {
            messages[index] = Message(
                id: messages[index].id,
                role: .assistant,
                content: streamingContent
            )
        }
    }

    /// Finalizes a streaming response when the stream ends.
    private func handleStreamEnd(messageId: String) {
        if let index = messages.lastIndex(where: { $0.id == streamingMessageId }) {
            messages[index] = Message(
                id: messageId,
                role: .assistant,
                content: streamingContent
            )
        }

        finalizeStreamingMessage()
    }

    /// Resets streaming state after a stream completes or is interrupted.
    private func finalizeStreamingMessage() {
        isStreaming = false
        streamingContent = ""
        streamingMessageId = nil
    }

    /// Converts any error into a user-friendly error message.
    private func handleError(_ error: Error) {
        self.error = userFriendlyMessage(for: error)
    }

    /// Returns a user-friendly message string for a given error.
    private func userFriendlyMessage(for error: Error) -> String {
        if let apiError = error as? APIError {
            return apiError.errorDescription ?? "Something went wrong. Please try again."
        }
        return "Something went wrong. Please try again."
    }
}

File: android/app/src/main/java/com/algonit/algo/features/chat/presentation/ChatViewModel.kt
package com.algonit.algo.features.chat.presentation

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.algonit.algo.core.network.WebSocketClient
import com.algonit.algo.core.network.WsEvent
import com.algonit.algo.core.util.AppResult
import com.algonit.algo.features.chat.data.model.ChatAPIResponse
import com.algonit.algo.features.chat.data.model.Message
import com.algonit.algo.features.chat.data.model.SuggestedAction
import com.algonit.algo.features.chat.domain.repository.ChatRepository
import com.algonit.algo.features.chat.domain.usecase.SendMessageUseCase
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * UI state for the chat screen.
 */
data class ChatUiState(
    val messages: List<Message> = emptyList(),
    val isLoading: Boolean = false,
    val isStreaming: Boolean = false,
    val streamingText: String = "",
    val error: String? = null,
    val suggestedActions: List<SuggestedAction>? = null,
    val conversationId: String? = null,
    val hasMoreMessages: Boolean = false,
    val isLoadingMore: Boolean = false
)

/**
 * ViewModel managing the chat conversation state.
 * Handles message sending, WebSocket streaming, action confirmation,
 * and message history pagination.
 */
@HiltViewModel
class ChatViewModel @Inject constructor(
    private val sendMessageUseCase: SendMessageUseCase,
    private val chatRepository: ChatRepository,
    private val webSocketClient: WebSocketClient,
    savedStateHandle: SavedStateHandle
) : ViewModel() {

    private val _uiState = MutableStateFlow(ChatUiState())
    val uiState: StateFlow<ChatUiState> = _uiState.asStateFlow()

    // Public convenience accessors for Compose
    val messages: StateFlow<List<Message>> get() = MutableStateFlow(_uiState.value.messages).also { flow ->
        viewModelScope.launch {
            _uiState.collect { (flow as MutableStateFlow).value = it.messages }
        }
    }

    val suggestedActions: StateFlow<List<SuggestedAction>?> get() = MutableStateFlow(_uiState.value.suggestedActions).also { flow ->
        viewModelScope.launch {
            _uiState.collect { (flow as MutableStateFlow).value = it.suggestedActions }
        }
    }

    private var currentPage = 1

    init {
        // Restore conversation ID if navigating back
        val convId = savedStateHandle.get<String>("conversationId")
        if (convId != null) {
            _uiState.update { it.copy(conversationId = convId) }
            loadMessageHistory(convId)
        }

        // Subscribe to WebSocket events for streaming
        observeWebSocketEvents()

        // Connect WebSocket
        webSocketClient.connect()
    }

    /**
     * Sends a text message to the assistant.
     */
    fun sendMessage(text: String) {
        val trimmed = text.trim()
        if (trimmed.isBlank()) return

        viewModelScope.launch {
            // Add optimistic user message to UI
            val userMessage = Message.createUserMessage(trimmed)
            _uiState.update { state ->
                state.copy(
                    messages = state.messages + userMessage,
                    isLoading = true,
                    error = null,
                    suggestedActions = null
                )
            }

            // Send via use case
            when (val result = sendMessageUseCase(
                conversationId = _uiState.value.conversationId,
                message = trimmed
            )) {
                is AppResult.Success -> {
                    handleChatResponse(result.data)
                }
                is AppResult.Error -> {
                    _uiState.update { state ->
                        state.copy(
                            isLoading = false,
                            error = result.error.toUserMessage()
                        )
                    }
                }
            }
        }
    }

    /**
     * Sends a voice-transcribed message.
     */
    fun sendVoiceMessage(text: String) {
        val trimmed = text.trim()
        if (trimmed.isBlank()) return

        viewModelScope.launch {
            val userMessage = Message.createUserMessage(trimmed, inputType = "voice")
            _uiState.update { state ->
                state.copy(
                    messages = state.messages + userMessage,
                    isLoading = true,
                    error = null,
                    suggestedActions = null
                )
            }

            when (val result = sendMessageUseCase(
                conversationId = _uiState.value.conversationId,
                message = trimmed,
                inputType = "voice"
            )) {
                is AppResult.Success -> handleChatResponse(result.data)
                is AppResult.Error -> {
                    _uiState.update { it.copy(isLoading = false, error = result.error.toUserMessage()) }
                }
            }
        }
    }

    /**
     * Executes a suggested action by sending it as a message.
     */
    fun executeSuggestedAction(action: SuggestedAction) {
        sendMessage(action.label)
    }

    /**
     * Confirms or cancels a pending action.
     */
    fun confirmAction(confirmationId: String, confirmed: Boolean) {
        val conversationId = _uiState.value.conversationId ?: return

        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }

            when (val result = chatRepository.confirmAction(
                conversationId = conversationId,
                confirmationId = confirmationId,
                confirmed = confirmed
            )) {
                is AppResult.Success -> handleChatResponse(result.data)
                is AppResult.Error -> {
                    _uiState.update { it.copy(isLoading = false, error = result.error.toUserMessage()) }
                }
            }
        }
    }

    /**
     * Loads the next page of message history.
     */
    fun loadMoreMessages() {
        val conversationId = _uiState.value.conversationId ?: return
        if (_uiState.value.isLoadingMore || !_uiState.value.hasMoreMessages) return

        viewModelScope.launch {
            _uiState.update { it.copy(isLoadingMore = true) }
            currentPage++

            when (val result = chatRepository.getMessages(conversationId, currentPage)) {
                is AppResult.Success -> {
                    val olderMessages = result.data
                    _uiState.update { state ->
                        state.copy(
                            messages = olderMessages + state.messages,
                            isLoadingMore = false,
                            hasMoreMessages = olderMessages.isNotEmpty()
                        )
                    }
                }
                is AppResult.Error -> {
                    currentPage--
                    _uiState.update { it.copy(isLoadingMore = false) }
                }
            }
        }
    }

    /**
     * Clears the current error.
     */
    fun clearError() {
        _uiState.update { it.copy(error = null) }
    }

    /**
     * Processes a successful chat API response.
     */
    private fun handleChatResponse(response: ChatAPIResponse) {
        val assistantMessage = response.toMessage()

        _uiState.update { state ->
            state.copy(
                messages = state.messages + assistantMessage,
                isLoading = false,
                isStreaming = false,
                streamingText = "",
                conversationId = response.conversationId,
                suggestedActions = response.response.suggestedActions
            )
        }
    }

    /**
     * Loads message history for an existing conversation.
     */
    private fun loadMessageHistory(conversationId: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true) }

            when (val result = chatRepository.getMessages(conversationId)) {
                is AppResult.Success -> {
                    _uiState.update { state ->
                        state.copy(
                            messages = result.data,
                            isLoading = false,
                            hasMoreMessages = result.data.size >= 50,
                            suggestedActions = result.data.lastOrNull()?.suggestedActions
                        )
                    }
                }
                is AppResult.Error -> {
                    _uiState.update { it.copy(isLoading = false, error = result.error.toUserMessage()) }
                }
            }
        }
    }

    /**
     * Subscribes to WebSocket events for real-time streaming responses.
     */
    private fun observeWebSocketEvents() {
        viewModelScope.launch {
            webSocketClient.events.collect { event ->
                handleWebSocketEvent(event)
            }
        }
    }

    /**
     * Handles individual WebSocket events for streaming and alerts.
     */
    private fun handleWebSocketEvent(event: WsEvent) {
        val currentConvId = _uiState.value.conversationId

        when (event.type) {
            "typing" -> {
                if (event.conversationId == currentConvId) {
                    _uiState.update { it.copy(isStreaming = true, streamingText = "") }
                }
            }
            "stream" -> {
                if (event.conversationId == currentConvId) {
                    _uiState.update { state ->
                        state.copy(
                            isStreaming = true,
                            streamingText = state.streamingText + (event.chunk ?: "")
                        )
                    }
                }
            }
            "stream_end" -> {
                if (event.conversationId == currentConvId) {
                    _uiState.update { it.copy(isStreaming = false) }
                }
            }
        }
    }

    override fun onCleared() {
        super.onCleared()
        webSocketClient.disconnect()
    }
}
